bundles/org.openhab.binding.plugwiseha/src/main/java/org/openhab/binding/plugwiseha/internal/handler/PlugwiseHAApplianceHandler.java

```diff
@@ -0,0 +1,430 @@
+                }
+                break;
+            default:
+                break;
+        }
+
+        if (state != UnDefType.NULL && state != UnDefType.UNDEF) {
+            updateState(channelID, state);
+        }
+    }
+
+    protected synchronized void addBatteryChannels() {
+        logger.debug("Battery operated appliance detected: adding 'Battery level' and 'Battery low level' channels");
+
+        ChannelUID channelUIDBatteryLevel = new ChannelUID(getThing().getUID(), APPLIANCE_BATTERYLEVEL_CHANNEL);
+        ChannelUID channelUIDBatteryLevelLow = new ChannelUID(getThing().getUID(), APPLIANCE_BATTERYLEVELLOW_CHANNEL);
+
+        boolean channelBatteryLevelExists = false;
+        boolean channelBatteryLowExists = false;
+
+        List<Channel> channels = getThing().getChannels();
+        for (Channel channel : channels) {
+            if (channel.getUID().equals(channelUIDBatteryLevel)) {
+                channelBatteryLevelExists = true;
+            } else if (channel.getUID().equals(channelUIDBatteryLevelLow)) {
+                channelBatteryLowExists = true;
+            }
+            if (channelBatteryLevelExists && channelBatteryLowExists) {
+                break;
+            }
+        }
+
+        if (!channelBatteryLevelExists) {
+            ThingBuilder thingBuilder = editThing();
+
+            Channel channelBatteryLevel = ChannelBuilder.create(channelUIDBatteryLevel, "Number")
+                    .withType(CHANNEL_TYPE_BATTERYLEVEL).withKind(ChannelKind.STATE).withLabel("Battery level")
+                    .withDescription("Represents the battery level as a percentage (0-100%)").build();
+
+            thingBuilder.withChannel(channelBatteryLevel);
+
+            updateThing(thingBuilder.build());
+        }
+
+        if (!channelBatteryLowExists) {
+            ThingBuilder thingBuilder = editThing();
+
+            Channel channelBatteryLow = ChannelBuilder.create(channelUIDBatteryLevelLow, "Switch:Battery")
+                    .withType(CHANNEL_TYPE_BATTERYLEVELLOW).withKind(ChannelKind.STATE).withLabel("Battery low level")
+                    .withDescription("Switches ON when battery level gets below threshold level").build();
+
+            thingBuilder.withChannel(channelBatteryLow);
+
+            updateThing(thingBuilder.build());
+        }
+    }
+
+    protected void setApplianceProperties() {
+        Map<String, String> properties = editProperties();
+        logger.debug("Setting thing properties");
+
+        properties.put(PlugwiseHABindingConstants.APPLIANCE_PROPERTY_DESCRIPTION, this.appliance.getDescription());
+        properties.put(PlugwiseHABindingConstants.APPLIANCE_PROPERTY_TYPE, this.appliance.getType());
+        properties.put(PlugwiseHABindingConstants.APPLIANCE_PROPERTY_FUNCTIONALITIES,
+                this.appliance.getActuatorFunctionalities().keySet().stream().map(e -> e.toString())
+                        .collect(Collectors.joining(", ")));
```

Comment:
The keyset already seems to be a String. Why do the toString()?  This can also be done with join method:
```suggestion
                String.join(", ", this.appliance.getActuatorFunctionalities().keySet());
```